src/main/java/org/opensearch/knn/index/KNNWeight.java

```diff
@@ -117,12 +124,34 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
             final KNNQueryResult[] results;
             KNNCounter.GRAPH_QUERY_REQUESTS.increment();
 
+            // We need to first get index allocation
+            NativeMemoryAllocation indexAllocation = null;
+
+            try {
+                indexAllocation = nativeMemoryCacheManager.get(
+                        new NativeMemoryEntryContext.IndexEntryContext(
+                                indexPath.toString(),
+                                NativeMemoryLoadStrategy.IndexLoadStrategy.getInstance(),
+                                ImmutableMap.of(SPACE_TYPE, spaceType.getValue()),
+                                knnQuery.getIndexName()
+                        ), true);
+            } catch (ExecutionException e) {
+                throw new RuntimeException(e);
+            }
+
+            // Now that we have the allocation, we need to readLock it
+            indexAllocation.readLock();
```

Comment:
Why do we need read lock here? It is possible another search thread ends up here and ready to read this segment but because of this lock it potentially ends up blocking here?

Reply:
Many threads can obtain readlock as long as writelock is not held. I believe it is called many readers, one writer. An allocation cannot be read and written to at the same time: either it can be read by multiple threads or written by one thread. We grab the readlock so that no writelocks can be obtained at this point and modify our allocation.

Reply:
I think we need to lock it as soon as we get it. We  need to check isClosed before we query and after we lock it just in case it just got closed while we were waiting to get the lock.